E0308 makes incorrect proposal on range reference, missing parantheses

Hi there!

With this code (here's a [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=bc5a5b9495b5dc550dd38cac2262c565))
```rust
type Range = std::ops::Range<usize>;

fn demo(r: &Range) {
    println!("{:?}", r);
}

fn tell(x: usize) -> usize {
    x
}

fn main() {
    demo(tell(1)..tell(10)); // here, it is reported wrongly!
    demo(1..10); // here, it is reported correctly!
}
```

The compiler reports this error message with a hint
```
error[E0308]: mismatched types
  --> src/main.rs:13:10
   |
13 |     demo(tell(1)..tell(10)); // here, it is reported wrongly!
   |          ^^^^^^^^^^^^^^^^^
   |          |
   |          expected reference, found struct `std::ops::Range`
   |          help: consider borrowing here: `&tell(1)..tell(10)`
   |
   = note: expected reference `&std::ops::Range<usize>`
                 found struct `std::ops::Range<usize>`
```
but the hint is incorrect due to missing parantheses in this case. The correct answer would be
```
   |          help: consider borrowing here: `&(tell(1)..tell(10))`
```
It works correctly when the second line in main is executed, where the range is provided directly.

### Meta
`rustc --version --verbose`:
```
1.44.1
```

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->
This issue has been assigned to @ayazhafiz via [this comment](https://github.com/rust-lang/rust/issues/73553#issuecomment-647593388).
<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"ayazhafiz"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/src/librustc_hir/hir.rs
+++ b/src/librustc_hir/hir.rs
@@ -1511,13 +1511,7 @@ pub fn is_range_literal(sm: &SourceMap, expr: &Expr<'_>) -> bool {
     // Check whether a span corresponding to a range expression is a
     // range literal, rather than an explicit struct or `new()` call.
     fn is_lit(sm: &SourceMap, span: &Span) -> bool {
-        let end_point = sm.end_point(*span);
-
-        if let Ok(end_string) = sm.span_to_snippet(end_point) {
-            !(end_string.ends_with('}') || end_string.ends_with(')'))
-        } else {
-            false
-        }
+        sm.span_to_snippet(*span).map(|range_src| range_src.contains("..")).unwrap_or(false)
     };
 
     match expr.kind {

--- a/src/test/ui/range/issue-73553-misinterp-range-literal.rs
+++ b/src/test/ui/range/issue-73553-misinterp-range-literal.rs
@@ -1,0 +1,16 @@
+type Range = std::ops::Range<usize>;
+
+fn demo(r: &Range) {
+    println!("{:?}", r);
+}
+
+fn tell(x: usize) -> usize {
+    x
+}
+
+fn main() {
+    demo(tell(1)..tell(10));
+    //~^ ERROR mismatched types
+    demo(1..10);
+    //~^ ERROR mismatched types
+}

--- a/src/test/ui/range/issue-73553-misinterp-range-literal.stderr
+++ b/src/test/ui/range/issue-73553-misinterp-range-literal.stderr
@@ -1,0 +1,27 @@
+error[E0308]: mismatched types
+  --> $DIR/issue-73553-misinterp-range-literal.rs:12:10
+   |
+LL |     demo(tell(1)..tell(10));
+   |          ^^^^^^^^^^^^^^^^^
+   |          |
+   |          expected reference, found struct `std::ops::Range`
+   |          help: consider borrowing here: `&(tell(1)..tell(10))`
+   |
+   = note: expected reference `&std::ops::Range<usize>`
+                 found struct `std::ops::Range<usize>`
+
+error[E0308]: mismatched types
+  --> $DIR/issue-73553-misinterp-range-literal.rs:14:10
+   |
+LL |     demo(1..10);
+   |          ^^^^^
+   |          |
+   |          expected reference, found struct `std::ops::Range`
+   |          help: consider borrowing here: `&(1..10)`
+   |
+   = note: expected reference `&std::ops::Range<usize>`
+                 found struct `std::ops::Range<{integer}>`
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0308`.

